apps/ios/SimpleX NSE/NotificationService.swift

```diff
@@ -219,51 +266,82 @@ class NotificationService: UNNotificationServiceExtension {
         deliverBestAttemptNtf()
     }
 
+    func addExpectedMessage(ntfMessage: UserNtfMessage) {
+        if let connEntity = ntfMessage.connEntity_,
+           let receiveEntityId = connEntity.id, ntfMessage.expectedMsg_ != nil {
+            let expectedMsgId = ntfMessage.expectedMsg_?.msgId
+            let receivedMsgId = ntfMessage.receivedMsg_?.msgId
+            logger.debug("NotificationService: addExpectedMessage: expectedMsgId = \(expectedMsgId ?? "nil", privacy: .public), receivedMsgId = \(receivedMsgId ?? "nil", privacy: .public)")
+            expectedMessages[receiveEntityId] = ExpectedMessage(
+                ntfMessage: ntfMessage,
+                receiveConnId: connEntity.conn.agentConnId,
+                expectedMsgId: expectedMsgId,
+                allowedGetNextAttempts: 3,
+                msgBestAttempt: (
+                    ntfMessage.ntfsEnabled
+                    ? .nseData(.connectionEvent(ntfMessage.user, connEntity))
+                    : .empty
+                ),
+                ready: false
+            )
+        }
+    }
+
     override func serviceExtensionTimeWillExpire() {
         logger.debug("DEBUGGING: NotificationService.serviceExtensionTimeWillExpire")
         deliverBestAttemptNtf(urgent: true)
     }
 
-    func processReceivedNtf(_ ntf: NSENotification) -> Bool {
-        guard let ntfInfo = notificationInfo, let expectedMsgTs = ntfInfo.expectedMsg_?.msgTs else { return false }
+    func processReceivedNtf(_ id: ChatId, _ ntf: NSENotification) -> Bool {
+        guard
+            let expectedMessage = expectedMessages[id],
+            let expectedMsgTs = expectedMessage.ntfMessage.expectedMsg_?.msgTs
+        else {
+            return !expectedMessages.allSatisfy { $0.value.ready }
+        }
+        let ntfInfo = expectedMessage.ntfMessage
         if !ntfInfo.user.showNotifications {
-            self.setBestAttemptNtf(.empty)
+            expectedMessages[id]?.msgBestAttempt = .empty
```

Comment:
Currently just returning `!expectedMessages.allSatisfy { $0.value.ready }`, which means message loop continues while there are not ready "expected message" entries, and that messages previously returning false do not get processed by other threads if they exist. Simply returning false is not correct because then if there are no other matching threads loop will be stuck on waiting for a single message, and batch of expected messages would not finish processing (or would take longer than necessary).

Couple of options:
- We could change return to some enum instead of bool signalling "go onto next", "limited retry" (e.g. when not all expected messages are ready - try once or a few more times to find another thread, but shorter than current wait limit), "wait" (when all expected messages are ready - ok to wait as before).
- We could put these rejected messages to some other queue where they would be waiting for matching threads as before, without holding processing of this batch of expected messages.